src/linux/InotifyTree.cpp

```diff
@@ -109,13 +109,24 @@ void InotifyTree::removeNodeReferenceByWD(int wd) {
   }
 }
 
-void InotifyTree::renameDirectory(int wd, std::string oldName, std::string newName) {
-  auto nodeIterator = mInotifyNodeByWatchDescriptor->find(wd);
-  if (nodeIterator == mInotifyNodeByWatchDescriptor->end()) {
+void InotifyTree::renameDirectory(int oldDir, int newDir, std::string oldName, std::string newName) {
+  auto oldNodeIterator = mInotifyNodeByWatchDescriptor->find(oldDir);
+  if (oldNodeIterator == mInotifyNodeByWatchDescriptor->end()) {
+    return;
+  }
+
+  if (oldDir == newDir) {
+    oldNodeIterator->second->renameChild(oldName, newName);
+    return;
+  }
+
+  auto newNodeIterator = mInotifyNodeByWatchDescriptor->find(newDir);
+  if (newNodeIterator == mInotifyNodeByWatchDescriptor->end()) {
     return;
   }
 
-  nodeIterator->second->renameChild(oldName, newName);
+  oldNodeIterator->second->removeChild(oldName);
```

Comment:
I was just thinking about the performance implications of my approach, so if you've a better one I am good with it :) as far as I can see, your PR and mine are pretty much the same, but yours is more complete. The only thing worth adding to yours from here is the fix for moving outside of the watched directory, my last commit in this branch. Basically I check right after the inner loop finished if isGood is true (that I renamed to isStarted), if it is, then it means that there was a IN_MOVED_FROM without a IN_MOVED_TO, so the file/directory was moved outside of the watched tree, so it is a remove()

Reply:
I was actually tracking this exact issue this week while you were. I just didn't catch you on the feed :) https://github.com/Axosoft/nsfw/pull/63.

I added a test that caused the segfault in my PR, then fixed it in about the same way but with a I think a more performant tree maneuver.

Reply:
I'm glad to see this PR though, that means I was on the right track myself! Awesome timing lol :smile:

Reply:
I'll review a little further your implementation for out of watch tree movement and port it to my PR. Would that be good for you?

Reply:
Otherwise, if you'd like to reopen a PR with just that fix, I will merge it in so you have credit for finding that.

Reply:
Don't worry about credits, take the fastest path and just copy the code over, if the fix itself is good enough :) perhaps I'll eventually add a test for it later. Also because I'll have much time to code for the next few days unfortunately.

Reply:
I've cherrypicked your work on top of my branch. Thanks for your work!

Reply:
Thanks for this amazing piece of software!